Add uncaptured transactions count badge to Transactions menu

anu-rock opened the pull request:
Fixes #4504

#### Changes proposed in this Pull Request

Add a badge showing the number of pending authorizations (uncaptured transactions) in the Transactions entry in Payments WP menu.

#### Testing instructions

##### Pre-requisites:

* Enable the feature flag at `WC_Payments_Features::is_auth_and_capture_enabled`.
* Enable the setting `Issue an authorization on checkout, and capture later` [here](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=woocommerce_payments).
* Set your local WCPay server instance to listen to webhooks: ./local/bin/listen-to-webhooks.sh.

We'll start with the assumption that you do not have existing authorizations cached in your local database.

##### Capture a payment:

* Purchase an order from WC frontend.
* Refresh WP Admin. You should see a red badge (with count as 1) in WP Admin → Payments → Transactions.
* Go to the order's detail page from WP Admin → WooCommerce → Orders. Capture the payment from the "Order actions" menu.
* Refresh WP Admin. The red badge at WP Admin → Payments → Transactions should no longer be visible.

##### Cancel an authorization:

* Purchase another order from WC frontend.
* Refresh WP Admin. You should see a red badge (with count as 1) in WP Admin → Payments → Transactions.
* Go to the order's detail page from WP Admin → WooCommerce → Orders. Cancel the authorization from the "Order actions" menu.
* Refresh WP Admin. The red badge at WP Admin → Payments → Transactions should no longer be visible.

##### Disable manual capture:

* Purchase one more order from WC frontend.
* Refresh WP Admin. You should see a red badge (with count as 1) at WP Admin → Payments → Transactions.
* Disable the setting `Issue an authorization on checkout, and capture later`.
* Refresh WP Admin. The red badge at WP Admin → Payments → Transactions should no longer be visible.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

anu-rock committed:
Add badge for uncaptured transactions count in Payments menu

This is a basic implementation without caching, which will come a bit later.

anu-rock committed:
Display badge only when manual capture is enabled

This will save us a tonne of unnecessary API calls to get authorizations count.

anu-rock committed:
Cache authorization summary data

We use Database_Cache to store authorization summary data (including
uncaptured count) as an option in database.

Justification for caching:

* Uncaptured transactions count is fetched via API when the Payments menu
is created in WP Admin, which is essentially on each page load of WP Admin.

* Not all merchants will receive multiple orders every minute.

anu-rock committed:
Invalidate auth summary cache based on webhook events

anu-rock committed:
Add changelog

mgascam commented on includes/admin/class-wc-payments-admin.php:
```diff
@@ -420,6 +429,9 @@ public function add_payments_menu() {
 		$this->add_menu_notification_badge();
 		$this->add_update_business_details_task();
 		$this->add_disputes_notification_badge();
+		if ( $this->wcpay_gateway->get_option( 'manual_capture' ) === 'yes' ) {
```
❤️

kalessil commented on includes/admin/class-wc-payments-admin.php:
```diff
@@ -420,6 +429,9 @@ public function add_payments_menu() {
 		$this->add_menu_notification_badge();
 		$this->add_update_business_details_task();
 		$this->add_disputes_notification_badge();
+		if ( $this->wcpay_gateway->get_option( 'manual_capture' ) === 'yes' ) {
```
From stylistic PoV, the if statement can be moved inside 'add_transactions_notification_badge'.

anu-rock replied:
IMO, that's a potential violation of the single responsibility principle. `add_disputes_notification_badge` should do exactly that - add the badge.

Besides, having the conditions here makes it clear early on that adding the badge is dependent on conditions.

Also, a little above in the same method we have more such conditional UI rendering.

Let me know if you still think it's better to move the conditions inside the method.

kalessil replied:
I think we can look into this later when we'll be dropping feature flags. Marking this as resolved.

kalessil commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -161,6 +161,12 @@ public function process( array $event_body ) {
 			case 'charge.expired':
 				$this->process_webhook_expired_authorization( $event_body );
 				break;
+			case 'charge.succeeded':
```
Are we sure that server forwarding those events?

anu-rock replied:
You are right, the server is not. I realized this yesterday and changed the events locally. I just pushed this change in 2ad8f36.

anu-rock committed:
Invalidate cache based on intent rather than charge events

This will be things consistent with server. Besides, charge suceeded and
captured events are not currently being forwarded to the plugin.

kalessil commented:
Also, we are hiding UI changes behind a feature flag. So we should show budges behind a feature flag as well.
We are not sure yet when exactly we'll make UI visible to everyone and badges without a flag will be pretty confusing.

anu-rock committed:
Remove duplicate badge format constant

anu-rock committed:
Add count badge to Uncaptured tab button

Without a proper support for badges in WP's TabPanel component,
implementing this barebones but functional approach.

anu-rock committed:
Merge branch 'develop' into add/4504-auth-count-badge

anu-rock committed:
Fix duplicate import that came with merge with develop

anu-rock committed:
Add tests for badge in Transactions menu item

anu-rock committed:
Remove count badge from Uncaptured tab button

We'll revisit this later when we have a better solution.

anu-rock committed:
Hide the badge under a feature flag

anu-rock commented:
@kalessil 

> Also, we are hiding UI changes behind a feature flag. So we should show budges behind a feature flag as well.

Good point! Fixed in 24b60ff.

kalessil commented on includes/admin/class-wc-payments-admin.php:
```diff
@@ -943,7 +947,32 @@ public function add_disputes_notification_badge() {
+		}
+
+		$uncaptured_transactions = $this->get_uncaptured_transactions_count();
+
```
Can we drop this empty line, please?

anu-rock replied:
I'm not entirely happy with this, but here you go ddad6fe :)

kalessil commented on tests/unit/admin/test-class-wc-payments-admin.php:
```diff
@@ -285,7 +285,7 @@ public function test_disputes_notification_badge_display() {
-		$expected_badge = sprintf( WC_Payments_Admin::DISPUTE_NOTIFICATION_BADGE_FORMAT, 4 );
+		$expected_badge = sprintf( WC_Payments_Admin::UNRESOLVED_NOTIFICATION_BADGE_FORMAT, 4 );
 
 		$this->assertEquals( 'Disputes' . $expected_badge, $item_names_by_urls[ $dispute_url ] );
```
Since we are touching this, can we use strict assert here?

anu-rock replied:
Fixed in 2090c06.

kalessil commented on tests/unit/admin/test-class-wc-payments-admin.php:
```diff
@@ -318,4 +318,88 @@ public function test_disputes_notification_badge_no_display() {
+		// The expected badge content should include 3 uncaptured transactions.
+		$expected_badge = sprintf( WC_Payments_Admin::UNRESOLVED_NOTIFICATION_BADGE_FORMAT, 3 );
+
+		$this->assertEquals( 'Transactions' . $expected_badge, $item_names_by_urls[ $transactions_url ] );
```
Also reminder about strict asserts.

anu-rock replied:
Fixed in 2090c06.

kalessil commented on tests/unit/admin/test-class-wc-payments-admin.php:
```diff
@@ -318,4 +318,88 @@ public function test_disputes_notification_badge_no_display() {
+		$item_names_by_urls     = wp_list_pluck( $submenu[ WC_Payments_Admin::PAYMENTS_SUBMENU_SLUG ], 0, 2 );
+		$transactions_menu_item = $item_names_by_urls['wc-admin&path=/payments/transactions'];
+
+		$this->assertEquals( 'Transactions', $transactions_menu_item );
```
Also reminder about strict asserts.

anu-rock replied:
Fixed in 2090c06.

kalessil commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -312,6 +318,29 @@ private function process_webhook_expired_authorization( $event_body ) {
+	 * Process webhook for a payment intent canceled event.
+	 *
+	 * @param array $event_body The event that triggered the webhook.
+	 */
```
`@return void`

anu-rock replied:
Fixed in 2090c06.

kalessil commented on includes/class-wc-payments-webhook-processing-service.php:
```diff
@@ -312,6 +318,29 @@ private function process_webhook_expired_authorization( $event_body ) {
+	 * Process webhook for a payment intent amount capturable updated event.
+	 *
+	 * @param array $event_body The event that triggered the webhook.
+	 */
```
`@return void`

anu-rock replied:
Fixed in 2090c06.

kalessil commented on includes/admin/class-wc-payments-admin.php:
```diff
@@ -943,7 +947,32 @@ public function add_disputes_notification_badge() {
+	/**
+	 * Adds a notification badge to the Payments > Transactions admin menu item to
+	 * indicate the number of transactions that need to be captured.
+	 */
```
`@return void`

anu-rock replied:
Fixed in 2090c06.

kalessil commented on includes/admin/class-wc-payments-admin.php:
```diff
@@ -943,7 +947,32 @@ public function add_disputes_notification_badge() {
+		}
+
+		foreach ( $submenu[ self::PAYMENTS_SUBMENU_SLUG ] as $index => $menu_item ) {
+			if ( 'wc-admin&path=/payments/transactions' === $menu_item[2] ) {
```
Is that the approach we use for other badges as well?

anu-rock replied:
Indeed. For disputes:

https://github.com/Automattic/woocommerce-payments/blob/ddad6fe576519aaafefd2ba9f54b948bea6fc12e/includes/admin/class-wc-payments-admin.php#L944-L945

kalessil replied:
Gotcha, thanks!

kalessil reviewed (commented):
From coding PoV and approach look good!
I left a few minor comments and looking forward to hearing feedback from other teammates.

anu-rock committed:
Minor refactors for stricter typing

anu-rock committed:
Remove empty line between variable and condition

anu-rock committed:
Fix tests breaking because of feature flag

anu-rock committed:
Update changelog

kalessil reviewed (approved):
LGTM!

mgascam reviewed (approved):
LGTM!

mgascam committed:
Merge branch 'develop' into add/4504-auth-count-badge

anu-rock committed:
Merge branch 'develop' into add/4504-auth-count-badge
